### ISSUE TITLE
[DX] Provide UI to define watchdog severity levels

### ISSUE THREAD

**Issue by opi** [backdrop/backdrop-issues#3326]
When reviewing https://github.com/backdrop/backdrop-issues/issues/3279, I need to manually edit the `system.core.json` config file to update the `watchdog_enabled_severity_levels` setting.

This should be possible from the web interface. Following PR provides a basic checkboxes form with watchdog severity levels.

![Screen Shot 2019-11-24 at 3 09 51 pm](https://user-images.githubusercontent.com/2423362/69489483-9b4f3c80-0ecc-11ea-91d9-f306bbeae31d.png)

----
~~PR by @opi https://github.com/backdrop/backdrop/pull/2328~~
PR by @klonos (based on @opi's): https://github.com/backdrop/backdrop/pull/3001

**Comment by opi** [comment]
Related issue https://github.com/backdrop/backdrop-issues/issues/3325

**Comment by klonos** [comment]
PR looks really good @opi 👍 ...just a missing period in the help text + a minor suggestion to add the "Which" at the beginning.

Very useful!

**Comment by klonos** [comment]
Just noting: `system.core.json` before:

```
"watchdog_enabled_severity_levels": [
        0,
        1,
        2,
        3,
        4,
        5,
        6
    ],
```

...and after:

```
"watchdog_enabled_severity_levels": {
        "1": "1",
        "2": "2",
        "3": "3",
        "4": "4",
        "5": "5",
        "6": "6",
        "7": "7",
        "8": "8"
    },
```

**Comment by klonos** [comment]
...so after further testing on my local and on the PR sandbox, I am moving this to NW because when saving the form, no mater what selection of checkboxes you've made, the "Emergency" severity gets deselected.

**Comment by klonos** [comment]
...possibly related #3327 ??

**Comment by opi** [comment]
PR updated, with new wording (thanks!) and a fix for saved values.

**Comment by jlfranklin** [comment]
The UI to manage `watchdog_enabled_severity_levels` was added to the Devel module.  See backdrop-contrib/devel#4 for details.  (It was added late in the ticket.)

**Comment by klonos** [comment]
...I still think that this deserves to be in core Configuration >> Development >> Logging and errors.

**Comment by klonos** [comment]
Thanks for addressing the config save issue @opi 👍 RTBC!

**Comment by jenlampton** [comment]
Since this setting changes the way `dblog` module behaves, the new field and config value should belong in that module instead of `system`. See where the setting for `Database log messages to keep` lives, and put this one right after that, maybe?

I've also recommended a minor UX improvement on the PR.

**Comment by jenlampton** [comment]
No movement here since 2018. Removing milestone, adding candidate tag.

**Comment by klonos** [comment]
@jenlampton can you please re-review and point out what remains to be done with this? Have all your concerns been addressed in the PR, or do we need to discuss this further?

**Comment by jenlampton** [comment]
@klonos it looks like there are two tiny things left to do:

1) change the form element label to `Type of messages to log`
2) remove the form element description

**Comment by klonos** [comment]
New PR that fixes those last remaining bits (based on @opi's - so credits are attributed to him 😉): https://github.com/backdrop/backdrop/pull/3001

@jenlampton care to do a final review (and close that other PR)?

**Comment by dragonbot** [comment]
code looks great!  testing now...

**Comment by jenlampton** [comment]
code looks great! testing now...

**Comment by jenlampton** [comment]
1) On the dblog page, there are two filters. One is for "Types" and one is for "Severity". On the admin page, in the label of this new setting, we use the word "Types" when we mean "Severity". Let's change the field label to `Severity of messages to log`.

2) Shouldn't the severity filter on the logs page only show the types of messages that are actually in the log? If, for example, I have selected 5 severities of messages on the settings page, shouldn't I only see 5 options on the dblog page? Or better yet...  the Type filter only shows types that are in the database. Maybe we should do the same with the Severity filter?  (Should this be a separate issue?)

**Comment by klonos** [comment]
@jenlampton 

1. I have updated the PR to use `Severity of messages to log` for label 👍 

2. As per discussions on Gitter, yes, separate issue(s):
  - The checkboxes bit will be solved with #3069 and/or children issues, like #1800.
  - > If, for example, I have selected 5 severities of messages on the settings page, shouldn't I only see 5 options on the dblog page?
    Nope. Limiting the list of options of the "Severity" select to only those specified in `admin/config/development/logging` will not work as expected, because messages of different severities may have been previously logged (ones that are no longer set in `admin/config/development/logging`).
  - > Or better yet... the Type filter only shows types that are in the database. Maybe we should do the same with the Severity filter? (Should this be a separate issue?)
    Yup. Limiting the list of options to only what has been logged in the db may work better; but follow-up.

**Comment by jenlampton** [comment]
Works like a charm, RTBC!

**Comment by klonos** [comment]
@jenlampton (and others) I just noticed that the respective setting in Devel has the following description:

> Only messages with the enabled severity levels will be logged. The "debug" and "deprecated" levels are recommended for developer environments, but not on production sites.

Do you think that the second bit could be useful here? ...it would match the "It is recommended that sites running on production environments do not display any errors." help text for the types of error. Let me know, and I'll quickly update the PR.

PS: also making a note to self to file a PR that removes this setting from Devel, if system.module >= 1.14.2

**Comment by klonos** [comment]
...as per discussion on Gitter, I've updated the PR, to add the following help text:

> The _debug_ and _deprecated_ severity levels are recommended for developer environments, but not on production sites.

Leaving the status as RTBC, since @quicksketch will have a final look anyway 😅

**Comment by opi** [comment]
Seems all good. Thanks everyone for finishing that

**Comment by quicksketch** [comment]
I wasn't entirely certain about this change initially. Though if others feel it should be available then it's fine with me. I think that since we added the "Deprecated" option and started using it all over the place, this setting is a lot more likely to be used than before.

I've merged https://github.com/backdrop/backdrop/pull/3001 into 1.x for 1.15.0. With it right around the corner I'm not going to cherry-pick this back into 1.14.x.
